### PR TITLE
CORDA-2914: Allow certificate directory to be a symlink

### PR DESCRIPTION
**Pull request by manojkrgupta**
…ate-rpc-ssl-settings'

👮🏻👮🏻👮🏻 !!!! DESCRIBE YOUR CHANGES HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you run the unit, integration and smoke tests as described [here](https://docs.corda.net/head/testing.html)?
- [ ] If you added public APIs, did you write the JavaDocs?
- [ ] If the changes are of interest to application developers, have you added them to the [changelog](https://github.com/corda/corda/blob/master/docs/source/changelog.rst), and potentially the [release notes](https://github.com/corda/corda/blob/master/docs/source/release-notes.rst)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/corda/corda/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/corda/corda/blob/master/CONTRIBUTING.md#developer-certificate-of-origin).

Thanks for your code, it's appreciated! :)

### CODE REVIEW

**Comment by manojkrgupta** [comment]
Hi @mkit 

Kindly help review this pull request.

thanks

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt [hunk 1]
```diff
@@ -30,7 +31,14 @@ fun loadOrCreateKeyStore(keyStoreFilePath: Path, storePassword: String): KeyStor
         keyStoreFilePath.read { keyStore.load(it, pass) }
     } else {
         keyStore.load(null, pass)
-        keyStoreFilePath.toAbsolutePath().parent?.createDirectories()
+        if (keyStoreFilePath.toAbsolutePath().parent != null){
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: Space before the `{`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt [hunk 2]
```diff
@@ -30,7 +31,14 @@ fun loadOrCreateKeyStore(keyStoreFilePath: Path, storePassword: String): KeyStor
...
+            val parentDir = keyStoreFilePath.toAbsolutePath().parent
+            if (Files.isSymbolicLink(parentDir)){
```
Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: Space before the {

Comment (shamsasari) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: This is the second time where we've had this pattern of checking if the `Path` is a symbolic link before trying to read it (the other is in `initLogging` in NodeStartup.kt. Can you add a helper extension method `Path.safeSymbolicRead` in PathUtils.kt which does this if-else (but obviously without the `createDirectories`)

Comment (manojkrgupta) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: Liked the idea. Is it fine if I rebase this branch from master ? and then push in the changes. At present am not able to build any corda.jar from this branch (not sure why). But, with these changes, I am able to build corda.jar from master branch. Hence, asking, if I rebase (take all master commits) this branch, should there be any issues ?

Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: Yes that should be fine. I suggest you merge master into your branch as the easiest way to incorporate any new changes.

Comment (manojkrgupta) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/KeyStoreUtilities.kt]: Hi Sham,

Have accommodated the review comments. 
- Space before {
- New helper extension method Path.safeSymbolicRead in PathUtils.kt
- Using the new extension method in both the references - NodeStartup.kt, KeyStoreUtilities.kt

Tested -- looks good.

Let me know if you find anything else.
thanks

**Comment by shamsasari** [comment]
Thanks @manojkrgupta. This there a JIRA ticket we can link this to?

**Comment by manojkrgupta** [comment]
The dev ticket is https://r3-cev.atlassian.net/browse/CORDA-2914 (ongoing)
The support ticket is https://r3-cev.atlassian.net/browse/SUP-605 (closed)

I was following some convention of naming the branch name with the ticket number.

thanks

**Pull request merged**